feat(sort-filter): search project by name

NicoSerranoP opened the pull request:
**Description**
I am activating the new feature to search projects by name using the search bar. Behind the scenes what I am doing is I am taking the `registryAddress` and the search bar input and sending it to the server to perform a search. The server fetches all the metadata objects and add them to the projects. Later on it filters them comparing the `search` value and the `metadata.name` value.

**Considerations**
This current implementation performs the search on the server because it avoids problems with React `<InfiniteLoading>` and `useinfiniteQuery` in the client side. These two use pagination for better data efficiency and I did not want to break that feature. 

**Changes made**

- [X] Implement search projects by name feature
- [X] Improve server fetching for performance
- [X] Housekeeping to delete console.log errors
- [X] Housekeeping to prevent hydration errors

@ctrlc03 @kittybest @crisgarner @0xmad what do you think? 😊

kittybest reviewed (commented):
Thanks a lot! Just have a comment

NicoSerranoP commented on packages/interface/src/components/SortFilter.tsx:
```diff
@@ -8,31 +8,42 @@ import type { OrderBy, SortOrder } from "~/features/filter/types";
-        onChange={async (sort) => {
-          const [order, sorting] = sort.split("_") as [OrderBy, SortOrder];
-
-          await setFilter({ orderBy: order, sortOrder: sorting }).catch();
-        }}
+        onChange={onChangeSortByDropdown}
```
I just moved the `onChangeSortByDropdown` code above to use it with `useCallback` to make it a bit more efficient 😊

kittybest replied:
My question is this sort only contains these two:
<img width="249" alt="image" src="https://github.com/user-attachments/assets/4fe03e49-f8cb-45a6-8a4b-d0bf07aa1e6b" />

There should be `time_asc` and `time_desc` as well, now there's only `options={["name_asc", "name_desc"]}`, and it's not actually working even if I change the sortBy value.
But of course we could put this another issue.

NicoSerranoP replied:
I just created another issue to address this problem: https://github.com/privacy-scaling-explorations/maci-platform/issues/600

NicoSerranoP commented:
> You can use `useInfiniteQuery` for pagination support and throttling or debouncing support with `useDeferredValue` to not send requests every time you type something in a search bar. Current solution is not optimized if we have a big number of projects and it sends the requests on every search input change.

I added the `useDeferredValue` in the `Projects.tsx` component. Is that okay? 😊

0xmad commented:
> > You can use `useInfiniteQuery` for pagination support and throttling or debouncing support with `useDeferredValue` to not send requests every time you type something in a search bar. Current solution is not optimized if we have a big number of projects and it sends the requests on every search input change.
> 
> I added the `useDeferredValue` in the `Projects.tsx` component. Is that okay? 😊

You still need to use `useInfiniteQuery` for pagination support. Also, there are some configuration for requests optimization.
Imagine if you have a big number of projects and they all starts with your search input value. You fetch them all but you only need one page of results.

NicoSerranoP commented on packages/interface/src/components/ImageUpload.tsx:
This changes fixes two console errors:

1. A not-found file in the div style
2. A ref error involving the `Form` component and the `ImageUpload` component

@0xmad I was not able to use the ref from `fowardRef` but I ended up using the internal one. It is working correctly and not giving any console errors

NicoSerranoP commented on packages/interface/src/components/SingleRoundHome.tsx:
Small change to prevent reloading the page when clicking the View Projects button. It uses the NextJS router so a reload is not executed

NicoSerranoP commented on packages/interface/src/layouts/BaseLayout.tsx:
This change fixes two console errors:

1. A not-found file error
2. A hydration error by setting up the `theme` in a specific `clientTheme` variable

NicoSerranoP commented on packages/interface/src/providers/index.tsx:
This change fixes a hydration error caused by Wagmi. The SSR option does not change any wallet logic

NicoSerranoP commented:
@kittybest @0xmad hi guys! 😊 

I added the fixes for your current comments and also pushed my housekeeping fixes to prevent console errors. Could you review the file changes? I added a comment to each file to explain a bit more what I was doing.

After your final approval I will squash the commits to merge this PR


Thank you for your help :)

0xmad commented on packages/interface/src/components/ImageUpload.tsx:
```diff
@@ -1,82 +1,94 @@
+import { PolymorphicRef } from "./ui";
+
 export interface IImageUploadProps extends ComponentProps<"img"> {
   name?: string;
   maxSize?: number;
```
Fix these warning from eslint

0xmad commented on packages/interface/src/components/ImageUpload.tsx:
```diff
@@ -1,82 +1,94 @@
+      maxSize = 1024 * 1024, // 1 MB
+      className,
+    }: IImageUploadProps,
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    ref: PolymorphicRef<ElementType>,
```
If you don't use this ref, you don't need to add it as second param

NicoSerranoP replied:
I tried but I am getting a console log error saying that `forwardRef` needs a function that receives two parameters and it is not possible to remove the second one (ref). What I did is change its name to `_` and removed the types

0xmad replied:
In this case, you need to use `ref`. There is no reason for `forwardRef` if you don't use `ref`.

0xmad commented on packages/interface/src/components/ImageUpload.tsx:
```diff
@@ -1,82 +1,94 @@
+        control={control}
+        name={name}
+        render={({ field: { value, onChange, ...field } }) => (
+          // eslint-disable-next-line jsx-a11y/no-static-element-interactions, jsx-a11y/click-events-have-key-events
```
div shouldn't have `onClick` handlers.

NicoSerranoP replied:
This div contains the upload image button and the div icon. I tried to change `<div>` to `<button>` but it didn't work (we cannot have nested buttons)

0xmad replied:
Why do you need to have div handler `onClick` and button `onClick`? How is it supposed to work?

0xmad commented on packages/interface/src/components/ImageUpload.tsx:
```diff
@@ -1,82 +1,94 @@
-            style={{
-              backgroundImage: `url("${select.data ?? value}")`,
+            className={clsx("relative cursor-pointer overflow-hidden", className)}
+            onClick={() => {
+              internalRef.current?.click();
             }}
```
It should be `useCallback` function

0xmad commented on packages/interface/src/components/ImageUpload.tsx:
```diff
@@ -1,82 +1,94 @@
+              accept="image/png, image/jpeg"
+              className="hidden"
+              type="file"
+              // value={value?.[name]}
```
Remove it

0xmad commented on packages/interface/src/components/ImageUpload.tsx:
```diff
@@ -1,82 +1,94 @@
+              className="hidden"
+              type="file"
+              // value={value?.[name]}
+              onChange={(event) => {
+                const [file] = event.target.files ?? [];
+                if (file) {
+                  select.mutate(file, {
+                    onSuccess: (objectUrl) => {
+                      onChange(objectUrl);
+                    },
+                  });
+                }
+              }}
```
`useCallback`

0xmad commented on packages/interface/src/utils/fetchProjects.ts:
```diff
@@ -89,3 +91,47 @@ export async function fetchApprovedProjects(registryAddress: string): Promise<IR
+
+  const recipients = await Promise.all(
+    response.map(async (request) => {
+      const metadata = (await fetchMetadata(request.metadataUrl)) as unknown as Application;
+      if (search !== "") {
+        const name = metadata.name.toLowerCase();
+        if (!name.includes(search.trim().toLowerCase())) {
+          return null;
+        }
+      }
+      return {
```
And don't use nested ifs
```suggestion
      const metadata = (await fetchMetadata(request.metadataUrl)) as unknown as Application;
  
      if (search !== "") {
        const name = metadata.name.toLowerCase();
        if (!name.includes(search.trim().toLowerCase())) {
          return null;
        }
      }
     
      return {
```

0xmad reviewed (changes requested):
@NicoSerranoP thanks, just some changes requested

kittybest commented on packages/interface/src/components/ImageUpload.tsx:
```diff
@@ -1,82 +1,94 @@
-        });
-        throw new Error("IMAGE_TOO_LARGE");
-      }
+export const ImageUpload = forwardRef(
```
curious why change the `ImageUpload` here which should not be related to the search?
I saw you mentioned 2 errors here:
> A not-found file in the div style
> A ref error involving the Form component and the ImageUpload component

if this error is not because of your changes, then maybe fix it in another PR?

NicoSerranoP replied:
I added a couple of housekeeping fixes 😊 (that is why I am modifying the `ImageUpload` component). I also updated the PR description with some of the things I worked on

kittybest commented on packages/interface/src/utils/fetchProjects.ts:
```diff
@@ -89,3 +91,45 @@ export async function fetchApprovedProjects(registryAddress: string): Promise<IR
+
+/**
+ * Fetch all approved projects with metadata
+ * @param registryAddress
```
there's another param called `search`

NicoSerranoP replied:
I just added the other parameter 😊 🚀

kittybest reviewed (commented):
Thanks for also the housekeeping! But maybe split the housekeeping part in another PR better in the future? make single PR only focus on single stuff.
For this PR is totally fine :)

NicoSerranoP commented:
The new PR is https://github.com/privacy-scaling-explorations/maci-platform/pull/603
